signimage/signimage.go

```diff
@@ -0,0 +1,357 @@
+		return authn.Anonymous, nil
+	}
+
+	f, err := os.Open(configfile)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	cf, err := config.LoadFromReader(f)
+	if err != nil {
+		return nil, err
+	}
+
+	var cfg dockertypes.AuthConfig
+
+	cfg, err = cf.GetAuthConfig(repo)
+	if err != nil {
+		return nil, err
+	}
+
+	return authn.FromConfig(authn.AuthConfig{
+		Username:      cfg.Username,
+		Password:      cfg.Password,
+		Auth:          cfg.Auth,
+		IdentityToken: cfg.IdentityToken,
+		RegistryToken: cfg.RegistryToken,
+	}), nil
+
+}
+
+func main() {
+	// get the env vars we are using for setup, or set some sensible defaults
+
+	var unsignedimagename string
+	var signedimagename string
+	var fileslist string
+	var privkeyfile string
+	var pubkeyfile string
+	var pullsecret string
+	var pushsecret string
+	var nopush bool
+	flag.StringVar(&unsignedimagename, "unsignedimage", "", "name of the image to sign")
+	flag.StringVar(&signedimagename, "signedimage", "", "name of the signed image to produce")
+	flag.StringVar(&fileslist, "filestosign", "", "colon seperated list of kmods to sign")
+	flag.StringVar(&privkeyfile, "key", "", "path to file containing private key for signing")
+	flag.StringVar(&pubkeyfile, "cert", "", "path to file containing public key for signing")
+	flag.StringVar(&pullsecret, "pullsecret", "", "path to file containing credentials for pulling images")
+	flag.StringVar(&pullsecret, "pushsecret", "", "path to file containing credentials for pushing images")
+	flag.BoolVar(&nopush, "no-push", false, "do not push the resulting image")
+
+	flag.Parse()
+
+	checkarg(&unsignedimagename, "UNSIGNEDIMAGE", "")
+	checkarg(&signedimagename, "SIGNEDIMAGE", unsignedimagename+"signed")
+	checkarg(&fileslist, "FILESTOSIGN", "")
+	checkarg(&privkeyfile, "KEYSECRET", "")
+	checkarg(&pubkeyfile, "CERTSECRET", "")
+	checkarg(&pullsecret, "PULLSECRET", "")
+	checkarg(&pushsecret, "PUSHSECRET", pullsecret)
+	// if we've made it this far the arguements are sane
+
+	// get a temp dir to copy kmods into for signing
+	extractiondir, err := os.MkdirTemp("/tmp/", "kmod_signer")
+	if err != nil {
+		fmt.Errorf("could not create temp dir: %v\n", err)
+		panic(err)
+	}
+
+	// sets up a tar archive we will use for a new layer
+	var b bytes.Buffer
+	tarwriter := tar.NewWriter(&b)
+
+	//make a map of the files to sign so we can track what we want to sign
+	kmodstosign := make(map[string]string)
+	for _, x := range strings.Split(fileslist, ":") {
+		kmodstosign[x] = "not found"
+	}
+
+	// set up image download otions
+	opts := make([]name.Option, 0)
+	ref, err := name.ParseReference(unsignedimagename, opts...)
+	if err != nil {
+		fmt.Errorf("could not parse the container image name: %v\n", err)
+		panic(err)
+	}
+
+	a, err := getAuthFromFile(pullsecret, strings.Split(unsignedimagename, "/")[0])
+	if err != nil {
+		fmt.Printf("failed to get auth: %v\n", err)
+		panic(err)
+	}
+
+	descriptor, err := remote.Get(ref, remote.WithAuth(a))
```

Comment:
We already have a registry package that has API for accessing the image, get layers of the image and checking if a file is present in the image. Maybe we can use it ( with some additional APIs). This will probably save a lot of code lines that you have here

Reply:
I disagree.

Firstly:  This is a standalone piece of code intended to run in a job that (may or may not be) run from KMMO. It is really only is in this repo at all because it need to be somewhere, and it not being in my personal github repo makes some external political things easier. The idea of closely coupling this code with that of the controller seems unnecessary. 

Secondly: After a quick look, while Registry superficially works in the same space, it would need significant changes to it make it usable, anything that references kmmv1beta1 would need to be removed, signimage does not have access to the crd, nor does it need it, but Registry refers to it extensively. 

Thirdly: Checking if a single image is present in the top layer is not enough here, we need to find all kmods (issue #94), in all layers (the code to loop through layers isn't in Registry), change them (not in Registry), add them to a new layer (not in Registry), write the image back (not in Registry). So there's a lot of functionality registry is missing we'd need to add.

So as it stands there is not much overlap between my code and Registry so it would need to be pretty much rewritten from scratch.

Reply:
actually, i think that the work to be done is not that big:
1) since your code is in this repo, why not the already available package? If, sometime in the future,  we decide that your code needs to be removed to a different repo, we can simply make registry a public package
2) registry is only using the PullOption struct of the kmmv1beta, it does not receive a CRD as a parameter anywhere. You will need to configure your access to the registry, so why not use kmmv1beta structs, as opposed to using structs defined in other packages?
3) we will deal with finding all kmod later ( your code right now also does not handle it). You main function can be describe as:

-  get the image
-  get all layers of image
-  loop over each layer and check if a file exist
-  in case file exists - get it, sign and add it
So we can add extraction function to the package and also add function, and use the existing function. I think is will significantly reduce your code and enhance our package, which is a good thing. The loop can stay in the main, just as it is not in preflight package, for example

Reply:
> we will deal with finding all kmod later ( your code right now also does not handle it).

yes it does, somewhere around line 284 
`if kmodstosign[canonfilename] == "not found" || (fileslist == "" && kmodstosign[canonfilename] == "") {`

this checks if the file has been passed in via filelist (that then populates kmodstosign),  or if filelist is empty and we haven't seen this kmod before, then sign anyway. 

(this is why Sign.FilesToSign in the api was set as optional in the crd :) )

I'll have a look at the rest of it tomorrow

Reply:
Lots of the generic code has been moved into internal/registry/registry.go and the binary has moved into cmd/signimage so tI think that resolves (or at least outdates) this